sockeye/inference.py

```diff
@@ -511,35 +535,151 @@ def get_max_output_length(input_length: int):
+    """
+    Returns a TranslatorInput object from a JSON object, serialized as a string.
+
+    :param sentence_id: An integer id.
+    :param json_string: A JSON object serialized as a string.
+    :return: A TranslatorInput.
+    """
+    jobj = json.loads(json_string, encoding=C.JSON_ENCODING)
+    tokens = jobj.get(C.JSON_TEXT_KEY)
+    if tokens is None:
+        raise ValueError("Missing 'text' field in JSON string: %s", json_string)
+
+    tokens = list(data_io.get_tokens(tokens))
+    factors = jobj.get(C.JSON_FACTORS_KEY)
+    if isinstance(factors, list):
+        factors = [list(data_io.get_tokens(factor)) for factor in factors]
+    else:
+        factors = None
+
+    return TranslatorInput(sentence_id=sentence_id, tokens=tokens, factors=factors)
+
+
+def make_input_from_factored_string(sentence_id: int,
+                                    factored_string: str,
+                                    translator: 'Translator',
+                                    delimiter: str = C.DEFAULT_FACTOR_DELIMITER) -> TranslatorInput:
+    """
+    Returns a TranslatorInput object from a string with factor annotations on a token level, separated by delimiter.
+    If translator does not require any source factors, the string is parsed as a plain token string.
+
+    :param sentence_id: An integer id.
+    :param factored_string: An input string with additional factors per token, separated by delimiter.
+    :param translator: A translator object.
+    :param delimiter: A factor delimiter. Default: '|'.
+    :return: A TranslatorInput.
+    """
+    utils.check_condition(bool(delimiter) and not delimiter.isspace(),
+                          "Factor delimiter can not be whitespace or empty.")
+    model_num_factors = translator.num_source_factors
+    num_additional_factors = model_num_factors - 1
+
+    if num_additional_factors == 0:
+        return make_input_from_plain_string(sentence_id=sentence_id, string=factored_string)
+
+    tokens = []  # type: Tokens
+    factors = [[] for _ in range(num_additional_factors)]  # type: List[Tokens]
+    for token_id, token in enumerate(data_io.get_tokens(factored_string)):
+        token, *token_factors = token.rsplit(delimiter, maxsplit=num_additional_factors)
+        utils.check_condition(bool(token), "Empty token at sentence {}, position {}".format(sentence_id, token_id))
+
+        token_factors = list(filter(None, token_factors))
+        utils.check_condition(len(token_factors) == num_additional_factors and all(token_factors),
+                              'Expecting {} factors, but got {} at sentence {}, word "{}"'.format(model_num_factors,
+                                                                                                  1+len(token_factors),
+                                                                                                  sentence_id,
+                                                                                                  token))
```

Comment:
I don't like this form of testing. It's a convoluted case, but assume we are expecting three factors (and assuming we removed the maxsplit before): 
`w1|f11|f12 w2|f21|f22`
for whatever reason the factor generation 1 doesn't work neither for w1 nor for w2, and factor generation 2 doesn't work for w1, and because of that it doesn't generate a space between the words. We end up with
`w1||w2||f22`
which is accepted with token="w1" and token_factors=["w2", "f22"]. Note that if maxsplit is kept, similar cases can also be constructed.  

I think not filtering and doing separate checks for lenght and not None would be more robust.

Reply:
I don't think I understand your example. We first split by whitespace, so you will not run into the case of no space between words, right? 
Did you have a look at the unit tests for this? Is there a case missing you think we do not handle properly?

Reply:
The example assumed empty factors resulting in None, which was not the case (see above). You can ignore this.

Reply:
Adding this test fails:
`("w1||w2||f22", 2, "|", 4, 1)`

Reply:
ok, I will see how to cover this